Use `WP_Term_Query` to fetch a post's terms.

boonebgorges opened the pull request:
See #2740.

The most parsimonious approach would've been to replace the `get_terms()` call with `new WP_Term_Query`, with the 'object_id' argument set to the value of`$request['post']`. But this would be a change in behavior for the term controller, since it'd mean bypassing the 'get_terms' filter. If we're sticking with the 'get_terms' filter, it seems consistent to use `wp_get_object_terms()` for object-specific queries, which is what I've done in this PR. Feedback welcome on this decision.

boonebgorges committed:
Use `WP_Term_Query` to fetch a post's terms.

Previously, when the 'post' parameter was passed to a terms collection
query, the `get_terms_for_post()` method was used, in order to leverage
the WP's taxonomy relationship cache. In WordPress 4.7, object-term
queries are natively cached, so the API-specific workaround can be
removed.

See #2740.

boonebgorges committed:
Simplify the counting of post terms.

boonebgorges commented:
Another note: this PR includes #2779, since the offending code addressed in that ticket is removed altogether here. Apologies for the crossed streams.

danielbachhuber commented:
> The most parsimonious approach would've been to replace the `get_terms()` call with new `WP_Term_Query`, with the `'object_id'` argument set to the value of `$request['post']`.

I think we should do whatever you think is best, and put our best foot forward. We've made breaking changes in the past, so I don't see that as a problem.

joehoyle commented:
@boonebgorges does this mean we no longer get the object caching on a post's terms?

boonebgorges commented:
> I think we should do whatever you think is best, and put our best foot forward. We've made breaking changes in the past, so I don't see that as a problem.

Thinking more about it: As a plugin author, I'd expect that my plugin filtering `'get_object_terms'` would affect all instances of object-term queries, API included. So the approach in the PR seems like the right one to me, even though it results in more code.

> does this mean we no longer get the object caching on a post's terms?

No. `wp_get_object_terms()` queries are cached starting in WP 4.7, since they now run through `WP_Term_Query`. (The caching mechanism is a bit different - it's the 'terms' cache group rather than `{$taxonomy}_relationships`.) See https://core.trac.wordpress.org/changeset/38667

I see that the PR is no longer applying properly because of #2779. I'll fix and resubmit.

boonebgorges committed:
Merge branch 'develop' into use-wp_term_query-cache

boonebgorges committed:
Fix variable name typo.

boonebgorges committed:
Add test demonstrating that object term queries are cached.

boonebgorges commented:
I've updated the PR:
- Fixed merge conflicts
- Fixed variable name typo
- Added test that demonstrates that the cache is being hit for object-term queries

The last item required a core bugfix, so the test will only pass against WP trunk https://core.trac.wordpress.org/changeset/38784

rmccue commented on tests/test-rest-tags-controller.php:
```diff
@@ -665,6 +665,31 @@ public function test_additional_field_update_errors() {
+		$found_2 = wp_list_pluck( $response->data, 'id' );
+
+		$this->assertEqualSets( $found_1, $found_2 );
+		$this->assertSame( $num_queries, $wpdb->num_queries );
```
This is currently failing in the tests.

kadamwhite replied:
@rmccue this is the test Boone indicated would only pass against trunk; are we using trunk in our test suite?

This patch looks good to me, other than I wonder if we can gate this test somehow for now.

joehoyle commented:
Hmm this is causing more issues, as the WP_Term_Query changes in 4.7 are making backwards compatibility for 4.6 break in several places. It's not just the new unit test that fails under 4.6, we have others ones too, which the change to using WP_Term_Query. I feel like we should be supported the latest release of WordPress, and right now that means supporting 4.6, so I'm -1 on merging this right now.

boonebgorges commented:
Tests were passing locally for me with WP nightly. Not sure why it's failing on Travis. I'll look into it more carefully.

> Hmm this is causing more issues, as the WP_Term_Query changes in 4.7 are making backwards compatibility for 4.6 break in several places. It's not just the new unit test that fails under 4.6, we have others ones too, which the change to using WP_Term_Query

@joehoyle When you say that "the change to using `WP_Term_Query`" is causing BC breaks, do you mean that (a) the change in the current PR breaks compatibility with sites running older versions of WP, or do you mean that (b) the change to `WP_Term_Query` in core is causing other BC breaks in the API?  If (b), please provide details. If (a), it's fine with me if the PR is not merged, if your rule is to support the current stable WP - but it should be part of any core merge, where BC with older versions of WP will not be an issue.

joehoyle commented:
@boonebgorges right, it's (b), sorry for not being clear. this isn't a backwards compat issue for core, it's because with this PR, we are switching old code to WP_Term_Query even under 4.6, but expecting it to do the same things as under 4.7, which isn't the case, of course.

It makes sense for this to be merged if we merge the API into core, as that code can then of course be dependant on 4.7, however as it stands right now, I'd rather not release another beta of the plugin a couple of days and break things for sites on 4.6 using the plugin.

joehoyle commented:
FYI tests _do_ pass on nightly.

boonebgorges commented:
Thanks, @joehoyle. I asked in Slack a couple days ago about your policy regarding fixes that depend on core patches, but didn't get a response :) Your policy sounds fine to me.

Tests don't seem to be passing on all versions of PHP. See eg https://travis-ci.org/WP-API/WP-API/builds/167090500. I don't understand why. There could be a race condition or something that's causing a cache miss (hard to believe that our tests are not bulletproof, I know)

joehoyle commented:
I think those tests were failing due to running on Nightly too early, I re-ran them and they all pass now.

boonebgorges commented:
I think that's called an "external race condition"

kadamwhite commented:
Marking this "blocked on core" since it depends on 4.7; maybe an over-extrapolation of the ticket but feels right :)

boonebgorges commented:
I've migrated to Trac. https://core.trac.wordpress.org/ticket/38504#ticket
